scripting/src/main/java/org/infinispan/scripting/impl/DistributedRunner.java

```diff
@@ -0,0 +1,45 @@
+package org.infinispan.scripting.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Future;
+
+import javax.script.ScriptException;
+
+import org.infinispan.Cache;
+import org.infinispan.commons.util.concurrent.Futures;
+import org.infinispan.commons.util.concurrent.NotifyingFuture;
+import org.infinispan.commons.util.concurrent.NotifyingFutureImpl;
+import org.infinispan.distexec.DefaultExecutorService;
+import org.infinispan.scripting.SystemBindings;
+import org.infinispan.scripting.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+public class DistributedRunner implements ScriptRunner {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass(), Log.class);
+   public static final DistributedRunner INSTANCE = new DistributedRunner();
+
+   private DistributedRunner() {
+   }
+
+   @Override
+   public <T> Future<T> runScript(ScriptingManagerImpl scriptManager, String scriptName, ScriptMetadata metadata, CacheScriptBindings binding) throws ScriptException {
+      Cache<?, ?> masterCacheNode = (Cache<?, ?>) binding.get(SystemBindings.CACHE);
+      if (masterCacheNode == null) {
+         throw log.distributedTaskNeedCacheInBinding(scriptName);
+      }
+      DefaultExecutorService des = new DefaultExecutorService(masterCacheNode);
+      List<Future<T>> tasks = des.submitEverywhere(new DistributedScript<T>(scriptName));
+      des.shutdown();
+      List<NotifyingFuture<T>> notifyingTasks = new ArrayList<>(tasks.size());
+      for (Future<T> task : tasks) {
+         NotifyingFutureImpl<T> nf = new NotifyingFutureImpl<>();
+         nf.setFuture(task);
+         notifyingTasks.add(nf);
+      }
+      NotifyingFuture<Void> combinedFutures = Futures.combine(notifyingTasks);
+
+      return null; // FIXME
```

Comment:
I have a bit of a design problem here: The signature of the ScriptRunner.runScript() method returns a Future<T>, where in reality we want to return a Future<List<T>>. In particular I would like to use @gustavonalle 's CompositeNotifyingFuture so that I return a future which completes only when all underlying futures have completed, aggregating the List<Future<T>> into a List<T> further wrapped in a future. The real question is: what common signature can I use for ScriptRunner.runtScript to satisfy both returning a single object and a list of objects ?